misc/labrea.1
.\"
.\" Copyright (c) 2002 Tom Liston <tliston@premmag.com>
.\"
.\" $Id: labrea.1,v 1.2 2003/09/12 21:23:39 lorgor Exp $
.\"
.TH LABREA 1
.SH NAME
labrea \- Honeypot for incoming IP connection attempts
.SH SYNOPSIS
\fBlabrea\fR [\fB-i --device INTERFACE\fR]
[\fB-n --network nnn.nnn.nnn.nnn[/nn]\fR]
[\fB-m --mask nnn.nnn.nnn.nnn\fR]
[\fB-t --throttle-size BYTES\fR]
[\fB-p --max-rate RATE \fR]
[\fB-R --soft-restart\fR]
[\fB-r --arp-timeout RATE\fR]
[\fB-s --switch-safe\fR]
[\fB-h --hard-capture\fR]
[\fB-x --disable-capture\fR]
[\fB-X --exclude-resolvable-ips\fR]
[\fB-P --persist-mode-only\fR]
[\fB-a --no-resp-synack\fR]
[\fB-H --auto-hard-capture\fR]
[\fB-f --no-resp-excluded-ports\fR]
[\fB--no-arp-sweep\fR]
[\fB--init-file FILE\fR]
[\fB-F --bpf-file FILE\fR]
[\fB-T --dry-run\fR]
[\fB-d --foreground\fR]
[\fB-o --log-to-stdout\fR]
[\fB-O --log-timestamp-epoch\fR]
[\fB-l --log-to-syslog\fR]
[\fB-b --log-bandwidth\fR]
[\fB-v --verbose\fR]
[\fB-q --quiet\fR]
[\fB-z --no-nag\fR]
[\fB-? --usage --help \fR]
[\fB-V --version\fR]
[\fB-I --ip-addr nnn.nnn.nnn.nnn\fR]
[\fB-E --my-mac-addr xx:xx:xx:xx:xx:xx\fR]
[\fB-D --list-interfaces\fR]
[\fB-j --winpcap-dev nn\fR]
[\fB--syslog-server nnn.nnn.nnn.nnn\fR]
[\fB--syslog-port nnn\fR]
.LP
[\fBBPF Filter\fR]
.SH DESCRIPTION
.B labrea
creates virtual machines for unused IP addresses in the specified
block of IP addresses.  LaBrea sits and listens for ARP "who-has"
requests.
.LP
When an ARP request for a particular IP goes unanswered for
longer than its "rate" setting (default: 3 seconds), labrea crafts an
ARP reply that routes all traffic destined for the IP to a "bogus" MAC
address.  labrea sniffs for TCP/IP traffic sent to that MAC
address and then responds to any SYN packet with a SYN/ACK packet that
it creates.
.SH OPTIONS
.B labrea
accepts the following options:
.TP
.BI "-i --device " interface
By default, labrea uses the first ethernet interface. This forces
labrea to use the specified interface.
.TP
.BR  "-n --network " xxx.xxx.xxx.xxx[/nn]
labrea normally pulls information about the netblock from the IP
information assigned to the interface.  If labrea is run on an
unconfigured interface (one without an assigned IP address), then use
this option to specify the subnet to be captured.
.RS
.LP
.I xxx.xxx.xxx.xxx
is the network address.
.I /nn
is the subnet mask in CIDR notation. If the subnet mask is not
specified here, then you must include the -m parameter.
.RE
.TP
.BR  "-m --mask " xxx.xxx.xxx.xxx
Another way to specify the network mask for the capture netblock. If
this parameter is specified, then the -n parameter must also be
specified.
.TP
.BR "-t --throttle-size " \fInn\fR
Sets the TCP window advertisement
to limit the amount of data sent to labrea. The number of data
bytes to allow per packet is 
.I nn
bytes.
.TP
.BR "-p --max-rate " \fIrate\fR
Connect attempts will be permanently captured by forcing the
connection into a "persist" state (by closing the TCP window). In this
state, the connection will not time out.  labrea will permanently
capture connect attempts up to maximum bandwidth
.I rate
bytes.  If the specified bandwidth is exceeded, labrea will still
tarpit the incoming connection (ie respond SYN/ACK to incoming SYN).
.TP
.BR "-R --soft-restart"
New captures will be held off for 5 minutes to let bandwidth
calculations progress. If a major scan hits just after startup, this
prevents labrea from capturing too many connections.
.TP
.BR "-r --arp-timeout " \fIrate\fR
Wait
.I rate
seconds after seeing incoming arp requests
before capturing an IP address.
.TP
.BR "-s --switch-safe"
When there is an incoming ARP request, specifies that labrea should
send out an ARP request of its own for the same IP address. This is
necessary for safe operation in a switched environment where one host
does not necessarily see all the traffic on the switch.
.TP
.BR "-h --hard-capture "
Once an IP address has been captured, then do not wait for a "-r"
timeout for the next incoming ARP request.
.TP
.BR "-x --disable-capture"
Do not capture IPs.
.TP
.BR "-X --exclude-resolvable-ips"
On startup, attempt DNS resolution on all IPs within
the capture netblock. Automatically exclude any IP that has a
corresponding entry in the DNS. Be careful because this can generate a
lot of DNS lookups if the capture subnet is large.
.TP
.BR "-P --persist-mode-only"
Try to limit bandwidth use by doing only persist capturing. Note: This
parameter has limited usefulness since below max b/w, the same
exchange that leads to persist capture also has the side effect
of tarpitting.
.TP
.BR "-a --no-resp-synack"
By default, the LaBrea virtual hosts respond to SYN/ACK with RST, and
answer Pings. Disables this behaviour.
.TP
.BR "-H --auto-hard-capture"
Mark all non-excluded and all non-hardexcluded IPs as being hard
captured. See 
.B labrea.conf(5)
for more information. This parameter should be used
.I with caution.
.TP
.BR "-f --no-resp-excluded-ports"
Drop incoming connections to excluded ports. Normal default behaviour
is to return a RST. Makes nmap-style scanning go much slower.
.TP
.BR "--no-arp-sweep"
On startup, labrea sweeps the capture subnet with bursts of ARP
requests in an attempt to locate all live machines. This parameter
disables the sweep.
.TP
.BR "--init-file " \fIfile\fR
Read the configuration from the specified
.I file
instead of from the default location.
.TP
.BR "-F --bpf-file " \fIfile\fR
Designates the name of a file containing a BPF filter pointing to
machines/ports to be tarpitted.  As with the command line BPF filter,
these connections MUST be firewalled to DROP inbound traffic.
.TP
.BR "-T --dry-run"
Do labrea initialization, including Dns excludes, parse of the
configuration file, opening the network interface etc. Print
diagnostic information, then exit.
.TP
.BR "-d --foreground"
Do not detach the process. (Unix systems only)
.TP
.BR "-o --log-to-stdout"
Send log information to stdout rather than to syslog.  This option
also implies and sets the -d option (i.e. do not detach process).
.TP
.BR "-O --log-timestamp-epoch"
Same as the "-o" option, but with time output in seconds since epoch
to make it easier for logfile analysis programs.
.TP
.BR "-l --log-to-syslog"
Send log messages to syslog.
.TP
.BR "-b --log-bandwidth"
Log a message every minute detailing the current bandwidth consumption
of the -p option (persist capture).
.TP
.BR "-v --verbose"
Increase the verbosity of log messages. Use twice for more effect.
.TP
.BR "-q --quiet"
Do not report arp requests for IPs that are not in the capture
subnet.
.TP
.BR "-z --no-nag"
Turn off the nag message. Before you do this, read the basic warning
in the Notes section just below.
.TP
.BR "-? --usage --help"
Print a help message and then exit.
.TP
.BR "-V --version"
Print version information and exit.
.TP
.BR "-I --ip-addr " \fInnn.nnn.nnn.nnn\fR
Manually specify the IP address for the labrea server.
.TP
.BR "-E --my-mac-addr " \fIxx:xx:xx:xx:xx:xx\fR
Manually specify the MAC address of the labrea server's NIC.
.TP
.BR "-D --list-interfaces"
On Windows systems, print the list of WinPcap devices, followed by the
list of the libdnet interfaces. Note that each API has a different
nomenclature for the underlying NIC.
.TP
.BR "-j --winpcap-dev " \fInn\fR
On windows systems, select the nth winpcap device in the list.
.SH NOTES
.SS Basic Warning about use of labrea
.I You must understand this:
As a default, LaBrea captures IP addresses by creating a "virtual
machine" that sits on any UNUSED IP address that it sees. labrea has
been carefully written and tested to transparently and peacefully
operate in normal production environments but ...
.LP
.I There is a potential for problems
if someone decides to start using one of the IP addresses that
labrea has laid claim to, or if labrea erroneously decides that an IP
address is free when in fact a real machine is already there.
.SS Built-in protections
labrea tries very hard to NEVER capture an IP that has a live machine
sitting on it.
.LP
The following automatic mechanisms are provided:
.RS
.IP \(bu 
If labrea sees a gratuitous ARP signalling the arrival of a new
machine, it marks the corresponding IP address as excluded. ("new kids
on the block" logic)
.IP \(bu
Each ARP response is noted and the corresponding IP address is
marked as excluded.
.IP \(bu
At startup, a systematic sweep is done of the entire capture subnet
(as long as the subnet is not too big). All IP addresses that respond
are marked as excluded.
.RE
.LP
Then there are ways of manually specifying the exclusion of certain
addresses, and otherwise ensuring safe operation:
.RS
.IP \(bu
The EXC config stmt allows specified IP addresses to be manually
excluded from capture.
.IP \(bu
The IPI config stmt causes packets with the specified IP source
address(es) to be ignored.
.IP \(bu
-s --switch-safe parameter causes mirroring of ARP requests in a
switched environment
.IP \(bu
-X --exclude-resolvable-ips says to exclude all IPs that have a
corresponding Dns entry
.RE
.LP
.B Traffic rerouting:
Despite all this, if labrea somehow receives traffic whose IP
destination address belongs to a live machine, labrea will reroute
that traffic to the real machine.
.SS Size of the capture subnet
.LP
It is best to limit the capture subnet to the
.I actual physical segment
(VLAN, hub) where labrea is running.
.LP
In some configurations, where
proxy arp is being used to determine routing, interface subnet masks
can be quite large. (i.e. the "whole" network is "directly" attached to
the physical segment). 
.LP
In this case, if labrea picks up the subnet mask from the interface,
then labrea will inefficiently watch addresses that it has no hope of
capturing. You should use the -m / -n parameters to manually limit the
size of the capture subnet.
.SS Other usage notes
.HP
The labrea virtual machines use a bogus MAC address of 0:0:f:ff:ff:ff
.HP
On certain older Windows systems, it may be necessary to manually
specify the capture subnet.
.HP
On unix systems, KILL -USR1 will toggle logging off on and off.
.HP
On unix systems, KILL -HUP will cause labrea to reinitialize (and thus
free captured IPs).
.HP
If the capture subnet is too large (greater than 1024 addresses), then
labrea will not do an arp sweep.
.SH BUGS
.HP
On some systems, if there is absolutely no traffic to sniff,
pcap_dispatch will wait instead of timing out, making the program seem
unresponsive. (Workaround: ping the labrea server to "wake" it up.)
.HP
If --exclude-resolvable-ips is enabled, and if the capture subnet is
large (say class A /8), then a LOT of traffic will be generated to the
Dns server.
.SH EXAMPLES
.IP 1)
Run safely in a switched environment with very verbose logging. Don't
respond to excluded ports. Log bandwidth usage from persist
capturing. Exclude all IPs that are in the Dns. Run in the foreground,
and log to stdout. Maximum capture bandwidth is 2 MB/sec. Use
toto.conf as the initialisation file. Use network device "eth1"
instead of the default device. Do a test run only - parse input,
initialize, then exit.
.LP
.nf
	  labrea --switch-safe --verbose -v  --no-resp-excluded-ports
		--log-bandwidth --exclude-resolvable-ips --foreground
		--log-to-stdout --max-rate 2000000 --init-file toto.conf
		--device eth1 -z --dry-run
		(one line)
.fi
.IP 2)
Same thing with the short parameter style.
.LP
.nf
	  labrea -z -s -v -v -f -b -X -d -o -p 2000000
		--init-file toto.conf -i eth1 -T
		(one line)
.fi
.SH FILES
.TP
.I /usr/local/etc/labrea.conf
Default configuration file
.TP
.I /usr/local/sbin/labrea
Program
.SH SEE ALSO
\fBlabrea.conf\fR(5)
.SH AUTHOR
Tom Liston <tliston@hackbusters.net>
Bugs: lorgor@users.sourceforge.net or http://labrea.sourceforge.net



NEWS
2.5-beta-2
	-This is the release where we changed the cmd line parameters
        (again -( ).
	
	-Unit tested on Win2K, Win98, linux, FreeBSD, Solaris

	-Add support for libdnet-1.7. Since libdnet's API is cleaner,
	lbio became simpler (good). But the "-i" parameter in windows
	now is standard, and the -j parameter now refers to the
	WinPcap device. Sorry about this change.

	-There was an integer overflow problem that had to be dealt
	with. At the same time, we decided to change --max-rate (-p)
	parameter to assume input is in units of Kbits to avoid
	dealing with large bandwidths expressed in bytes. Keep
	internal totals in KByte units. Set an upper limit on max-rate
	to avoid integer overflows during calculations. Upper limit is
	currently 1 GByte/sec = 8 Gbits/sec. Sorry if this change
	causes inconvenience.

	-Corrected problem with timer pop so that bandwidth usage will
	be reported. Certain OSs were not working properly.


2.5-beta-1
	- Unit tested on Win98, WinME, WinXP, linux, OpenBSD, Solaris

	- Added debugging support. Has to be compiled into the pgm via
          ./configure --enable-debug. Then is controlled with cmd line
          --debug parameter.

2.5-alpha-1
	- Add support for automake, autoconf

	- Make code depend on libdnet. Use libdnet defines throughout
          the code.

	- (dynamic firewall ports) If ports are firewalled (ie no
	RST), then activity above a certain threshold will make a port
	start to respond.

	- The firewall code now checks the dest port on all incoming
          TCP packets.

	- Filter changed to make pgm will only hear packets *sent* to
	the bogus MAC address instead of all packets.

	- Responses will seem to come from the bogus virtual machine
	instead of the actual MAC address of the labrea server.

	- (arp sweeps) Pgm does an arp sweep of the capture subnet to
	try to locate live machines. A new parameter was added to turn
	this behaviour off if desired. For the arp sweep to happen,
	the capture subnet must be a "reasonable" size. Arps are sent
	out in batches of 80 at a time, at 1 - 2 minute intervals.

	- Pgm now takes notes of replies to Arp WHO-HAS and will leave
          these addresses alone (ie "new kid").

	- The "new kids on the block" logic was converted to use an
          array instead of a linked list. This should speed up the pgm
          and simplify the logic at the expense of storage. The "new
          kids" culling logic was moved into the timer signal handler
          code so that the new kid list is run at regular intervals
          but not at each arp.

	- The persistent connection logic was changed (newthisminute)
          so that the decision is made based on b/w and not just
          number of connections.
	
	- Test mode now logs on sysout. Test mode will not fork a
	child.

	- Pgm accepts long options at invocation (ie
          "--my-option"). Usage function was modified to show the long
          options.

	- Pgm will attempt to parse all input before bailing out.

	- Pcap_dispatch is used instead of Pcap_loop. This makes the
	pgm more efficient but less responsive to signals on certain
	operating systems.

	- "-m" parameter is still supported but "-n" can understand a
	CIDR format IP address (ie xxx.xxx.xxx.xxx/nn)

	- Catastrophic execution error messages go to std
	error. (warnx)

	- Old style config files are no longer supported.

	- "PMN" directive added to config file to force a port to be
          monitored.

	- IPI config statement was changed to be in CIDR notation
          (xx.xx.xx.xx/nn). The IP ignore list is now a linked list of
          libdnet addr structures.

	- Pgm logic was restructured. Globals were mostly moved into
          static structures. Enums were used instead of defines. Some
          #ifdefs were replaced by conditional statements. Message
          strings were moved back into the main code. Externs and
          gotos were eliminated. queue.h support was used for linked
          lists.

	- Signal handling was changed to set a flag only. Signal
          processing logic is driven by this flag off the mainline
          loop.

	- Logging to syslog starts after the pgm's initialisation is
          completed.

	- Added utility print function to centralize all messages and
          logging.

	- bget functions used for dynamic memory allocation. Runs
          faster than native support on some operating systems.

	  *** Windows-specific ***

	- Windows version was developed on Cygwin / Mingw. Autoconf
          support was added for Windows.

	- Syslog support for Windows Event file as well as logging to
          a remote syslog daemon now works. If remote syslog doesn't
          work, then fails over to local event log. On Win98 / WinME,
          an attempt to log to the non-existent local event log causes
          an error message.

	- Remote syslog code (which was shamelessly borrowed from
          snort) was modified to open the socket (with possible
          accompanying DNS search) once at pgm startup, instead of for
          each new message.

	- A new parameter "-j" was added to select the WinPcap
          driver. The libdnet intf rtns support "-i" as is the case
          under Unix. The libdnet rtns are used to determine interface
          Mac and IP addresses.

	- A new parameter "-D" allows the user to obtain a display of
          all the libdnet and WinPcap devices. The display also shows
          the default devices.

	- On Windows, by default logging is to stdout, not syslog.


## $Id: NEWS,v 1.2 2003/09/09 20:49:24 lorgor Exp $


src/labrea.c
/*
 * LaBrea
 *
 * A program to create a tar pit for the connection attempts to a
 * block of ip addresses. LaBrea sits and listens for ARP "who-has"
 * requests.  When an ARP request for a particular IP goes unanswered
 * for longer than it's "rate" setting (default: 3 seconds), LaBrea
 * crafts an ARP reply that routes all traffic destined for the IP to
 * a "bogus" MAC address.  LaBrea then listens for TCP/IP traffic
 * routed to that MAC address and then responds to any SYN packet with
 * a SYN/ACK packet that it creates.
 *
 * LaBrea completely ignores any other input.  This, forces the
 * "client" to wait for its TCP stack to time-out the connection.
 * Because stacks tend to be a bit tenacious about "established"
 * connections, this should bog down the scanner for quite a long
 * while.
 *
 * THIS PROGRAM REQUIRES BOTH LIBDNET and LIBPCAP in order to compile.
 *
 * Copyright (C) 2001, 2002 Tom Liston <tliston@premmag.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 *
 * $Id: labrea.c,v 1.1 2003/01/09 18:13:19 lorgor Exp $ */

#include "config.h"

#ifdef HAVE_ERR_H
#include <err.h>
#else
#include "err.h"
#endif


#include "labrea.h"
#include "ctl.h"
#include "utils.h"
#include "lbio.h"
#include "pkt_handler.h"

/* Definitions of main control structures */
io_t io;
ctl_t ctl;
outp_t outp;

int
main (int argc, char **argv)
{
  labrea_init(argc, argv);	/* Do initialisation */

  util_set_signal_handlers();

  if (!WIN32_FLG)
    util_detach();	/* Fork to become a daemon if unix*/

  util_print(NORMAL,"Labrea started");

  /* loop! */
  for(;;) {
    if (pcap_dispatch(io.pcap, -1, &pkt_handler, (u_char *)&ctl) < 0) {
      util_print(NORMAL, "Error in pcap loop - EXITING: %s", pcap_geterr(io.pcap));
      util_clean_exit(1);
    }

    /* Handle signal(s) if set */
    if (ctl.signals) {
      if (ctl.signals & SIG_RESTART)
	util_restart();
      if (ctl.signals & SIG_QUIT)
	util_quit();
#ifndef WIN32
      if (ctl.signals & SIG_TIMER)
	util_timer();
      if (ctl.signals & SIG_TOGGLE_LOGGING)
	util_toggle_logging();
#endif
    }
  }
  util_clean_exit(0);
}
